File: pcr_sign.py
#!/usr/bin/env python3

import sys
import json
import base64
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec, utils
from cryptography.hazmat.primitives.serialization import load_pem_private_key, load_der_private_key

def generate_keypair():
    """
    Generates a P-384 (secp384r1) ECDSA keypair.
    Returns a tuple of (private_key, public_key).
    """
    private_key = ec.generate_private_key(ec.SECP384R1())
    public_key = private_key.public_key()
    return (private_key, public_key)

def export_keys_der_base64(private_key, public_key):
    """
    Exports keys in DER format, base64-encoded.
    Returns a tuple of (private_key_base64, public_key_base64).
    """
    # Export private key in PKCS#8 DER format
    private_der = private_key.private_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption()
    )
    
    # Export public key in SubjectPublicKeyInfo DER format
    public_der = public_key.public_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PublicFormat.SubjectPublicKeyInfo
    )
    
    # Base64 encode
    private_b64 = base64.b64encode(private_der).decode('ascii')
    public_b64 = base64.b64encode(public_der).decode('ascii')
    
    return (private_b64, public_b64)

def sign_data_p1363(private_key, data_bytes):
    """
    Signs `data_bytes` using ECDSA+SHA384, returning a 'raw' P1363 signature (base64).
    Steps:
      1. We do a normal ECDSA sign (which yields DER).
      2. We parse (r, s) from DER.
      3. We concatenate r|s in fixed 48-byte big-endian form.
      4. We base64-encode that raw signature for output.
    """
    # 1. DER-encoded signature
    der_sig = private_key.sign(
        data_bytes,
        ec.ECDSA(hashes.SHA384())
    )

    # 2. Decode DER to (r, s)
    (r_int, s_int) = utils.decode_dss_signature(der_sig)

    # For P-384: each of r, s must be 48 bytes (384 bits). We'll big-endian encode & pad to 48.
    r_bytes = r_int.to_bytes(48, byteorder='big')
    s_bytes = s_int.to_bytes(48, byteorder='big')

    # 3. Concatenate to form raw signature
    raw_sig = r_bytes + s_bytes

    # 4. Base64-encode
    return base64.b64encode(raw_sig).decode('ascii')


def main():
    # Example usage from the command line:
    #   ./pcr_sign.py generate-keys
    #   ./pcr_sign.py sign <base64-privkey> '{"PCR0":"...","timestamp":1234}'
    if len(sys.argv) < 2:
        print("Usage:")
        print("  pcr_sign.py generate-keys")
        print("  pcr_sign.py sign <base64-private-key> <json-string>")
        sys.exit(1)

    command = sys.argv[1]

    if command == "generate-keys":
        # Generate a new P-384 key pair, output them in base64 DER
        priv, pub = generate_keypair()
        priv_b64, pub_b64 = export_keys_der_base64(priv, pub)
        
        # Optionally also print them in PEM for human readability:
        private_pem = (
            "-----BEGIN PRIVATE KEY-----\n"
            + "\n".join(priv_b64[i:i+64] for i in range(0, len(priv_b64), 64))
            + "\n-----END PRIVATE KEY-----\n"
        )
        public_pem = (
            "-----BEGIN PUBLIC KEY-----\n"
            + "\n".join(pub_b64[i:i+64] for i in range(0, len(pub_b64), 64))
            + "\n-----END PUBLIC KEY-----\n"
        )
        
        # Display everything in a clear format
        print("===== KEYS FOR PCR SIGNING SYSTEM =====\n")
        
        print("===== FOR ENVIRONMENT VARIABLES =====")
        print(f"export SIGNING_PRIVATE_KEY='{priv_b64}'")
        print(f"export SIGNING_PUBLIC_KEY='{base64.b64encode(private_pem.encode()).decode()}'")
        print("")
        
        print("===== FOR FRONTEND INTEGRATION =====")
        print("// Base64-encoded DER public key for verifying PCR signatures (add to your frontend)")
        print(f"const PCR_VERIFICATION_PUBLIC_KEY_B64 = \"{pub_b64}\";")
        print("")
        
        print("===== ADDITIONAL KEY FORMATS =====")
        print("===== PRIVATE KEY (PKCS#8 DER BASE64) =====")
        print(priv_b64)
        print("\n===== PUBLIC KEY (SPKI DER BASE64) =====")
        print(pub_b64)
        print("\n===== PRIVATE KEY (PEM) =====\n" + private_pem)
        print("===== PUBLIC KEY (PEM) =====\n" + public_pem)
        
        print("\n===== VERIFICATION COMMAND =====")
        print("just verify-pcr-history dev")
        print("")

    elif command == "sign":
        if len(sys.argv) < 4:
            print("Usage: pcr_sign.py sign <base64-private-key> <json-string>")
            sys.exit(1)

        base64_priv = sys.argv[2]
        json_str = sys.argv[3]

        # Re-import the private key from base64 DER
        priv_der = base64.b64decode(base64_priv)
        private_key = serialization.load_der_private_key(priv_der, password=None)

        # Sign (raw p1363)
        sig_b64 = sign_data_p1363(private_key, json_str.encode('utf-8'))
        print("Signature (raw p1363, base64):", sig_b64)

    else:
        print("Unknown command:", command)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: pcr_sign.js
#!/usr/bin/env node

const fs = require('fs');
const crypto = require('crypto');

/**
 * Generates a P-384 (secp384r1) ECDSA keypair.
 * @returns {Object} An object containing the keypair information
 */
function generateKeypair() {
  // Generate an EC key pair on the P-384 curve
  const { publicKey, privateKey } = crypto.generateKeyPairSync('ec', {
    namedCurve: 'secp384r1',
  });

  // Export the private key in PKCS#8 DER format
  const privateKeyDer = privateKey.export({
    type: 'pkcs8',
    format: 'der',
  });

  // Export the public key in SPKI DER format
  const publicKeyDer = publicKey.export({
    type: 'spki',
    format: 'der',
  });

  // Base64-encode them for easy storage
  const privateKeyBase64 = privateKeyDer.toString('base64');
  const publicKeyBase64 = publicKeyDer.toString('base64');

  // Create PEM format for better human readability
  const privatePem = [
    '-----BEGIN PRIVATE KEY-----',
    ...privateKeyBase64.match(/.{1,64}/g) || [],
    '-----END PRIVATE KEY-----',
  ].join('\n');

  const publicPem = [
    '-----BEGIN PUBLIC KEY-----',
    ...publicKeyBase64.match(/.{1,64}/g) || [],
    '-----END PUBLIC KEY-----',
  ].join('\n');

  return {
    privateKeyBase64,
    publicKeyBase64,
    privatePem,
    publicPem,
    privateKey,
    publicKey
  };
}

/**
 * Signs data using ECDSA+SHA384 with P1363 (raw) encoding
 * @param {Buffer|string} privateKeyData - Private key in base64 DER format
 * @param {string} dataToSign - Data to sign
 * @returns {string} Base64-encoded signature
 */
function signData(privateKeyData, dataToSign) {
  // If privateKeyData is a string, assume it's base64 and convert to Buffer
  const privateKeyBuffer = Buffer.isBuffer(privateKeyData) 
    ? privateKeyData
    : Buffer.from(privateKeyData, 'base64');

  try {
    // Create the private key object
    const privateKey = crypto.createPrivateKey({
      key: privateKeyBuffer,
      format: 'der',
      type: 'pkcs8'
    });

    // Verify data is a string
    if (typeof dataToSign !== 'string') {
      throw new Error('Data to sign must be a string');
    }

    // Create the signer
    const signer = crypto.createSign('SHA384');
    signer.update(dataToSign);

    // Sign in "ieee-p1363" raw format which produces the raw r|s signature
    // This is directly compatible with Web Crypto API's verify method
    const signature = signer.sign({
      key: privateKey,
      dsaEncoding: 'ieee-p1363'
    });

    // Verify signature length (should be 96 bytes for P-384)
    if (signature.length !== 96) {
      console.warn(`Warning: Generated signature is ${signature.length} bytes (expected 96 for P-384)`);
    }

    return signature.toString('base64');
  } catch (error) {
    console.error("Error during signing:", error.message);
    throw error;
  }
}

/**
 * Main function to handle CLI commands
 */
function main() {
  const command = process.argv[2];

  if (!command) {
    console.log("Usage:");
    console.log("  pcr_sign.js generate-keys");
    console.log("  pcr_sign.js sign-pcr0 <pcr0-value>");
    process.exit(1);
  }

  if (command === "generate-keys") {
    const keys = generateKeypair();
    
    console.log("===== IMPORTANT: SAVE THESE KEYS SECURELY =====");
    console.log("Generate these keys once and securely store them for all future operations.");
    console.log("The private key is SENSITIVE and should be protected!\n");
    
    console.log("===== SET THESE ENVIRONMENT VARIABLES NOW =====");
    console.log(`export SIGNING_PRIVATE_KEY='${keys.privateKeyBase64}'`);
    console.log(`export SIGNING_PUBLIC_KEY='${keys.publicKeyBase64}'`);
    console.log("\nRun the above commands to set your environment variables.");
    console.log("Consider adding them to your .env file for persistence.\n");
    
    console.log("===== FOR FRONTEND INTEGRATION =====");
    console.log("// Add this constant to your frontend code for PCR signature verification");
    console.log(`const PCR_VERIFICATION_PUBLIC_KEY_B64 = "${keys.publicKeyBase64}";`);
    console.log("");
    
    console.log("===== KEY FORMATS (FOR REFERENCE) =====");
    console.log("PRIVATE KEY (PKCS#8 DER BASE64):");
    console.log(keys.privateKeyBase64);
    console.log("\nPUBLIC KEY (SPKI DER BASE64):");
    console.log(keys.publicKeyBase64);
    console.log("\nPRIVATE KEY (PEM):\n" + keys.privatePem);
    console.log("\nPUBLIC KEY (PEM):\n" + keys.publicPem);
    
    console.log("\n===== VERIFICATION COMMANDS =====");
    console.log("# Sign and append PCR measurements:");
    console.log("just append-pcr-dev");
    console.log("just append-pcr-prod");
    console.log("");
    console.log("# Verify PCR history signatures:");
    console.log("just verify-pcr-history dev");
    console.log("just verify-pcr-history prod");
    console.log("");
  }
  else if (command === "sign-pcr0") {
    // Get PCR0 value from argument or environment
    const pcr0 = process.argv[3];
    if (!pcr0) {
      console.error("Error: PCR0 value is required");
      console.log("Usage: pcr_sign.js sign-pcr0 <pcr0-value>");
      process.exit(1);
    }

    // Get private key from environment
    const privateKeyBase64 = process.env.SIGNING_PRIVATE_KEY;
    if (!privateKeyBase64) {
      console.error("Error: SIGNING_PRIVATE_KEY environment variable is not set");
      console.log("Set it with: export SIGNING_PRIVATE_KEY='your-base64-private-key'");
      process.exit(1);
    }

    try {
      // Sign just the PCR0 string
      const signature = signData(privateKeyBase64, pcr0);
      
      // Output only the signature (no comments or extra text)
      console.log(signature);
    } catch (error) {
      console.error("Error signing PCR0:", error.message);
      process.exit(1);
    }
  }
  else {
    console.log("Unknown command:", command);
    process.exit(1);
  }
}

// Run the main function if this script is executed directly
if (require.main === module) {
  main();
}

// Export functions for potential module usage
module.exports = {
  generateKeypair,
  signData
}; 